ext/opcache/shared_alloc_mmap.c

```diff
@@ -40,6 +40,350 @@
 # define MAP_HUGETLB MAP_ALIGNED_SUPER
 #endif
 
+#if defined(__x86_64__) && defined(__linux__)
+#define max(a,b) ((a) > (b) ? (a):(b))
+
+#if defined(MAP_HUGETLB)
+static const size_t huge_page_size = 2 * 1024 * 1024; /* 2MB page size */
+#else
+static const size_t ordinary_page_size = 4 * 1024;    /* 4KB page size */
+#endif /* MAP_HUGETLB */
+
+static const ptrdiff_t addr_4GB_distance = UINT32_MAX; /* 4GB address distance */
+static const uintptr_t addr_4GB = UINT32_MAX;          /* 4GB memory address */
+static const size_t    size_4GB = UINT32_MAX;          /* 4GB memory size */
+
+/* Store the segment address information */
+struct map_seg_addresses {
+    uintptr_t start; /* segment start address (included) */
+    uintptr_t end;   /* segment end address (excluded) */
+};
+
+/* Count segments in /proc/self/maps file; each line represents one segment. */
+static size_t map_segment_count(void)
+{
+    FILE *f = NULL;
+    f = fopen("/proc/self/maps", "r");
+    if (!f) return 0;
+
+    size_t count = 0; /* segment count */
+    char buffer[MAXPATHLEN];
+    while ((fgets(buffer, MAXPATHLEN, f)) != NULL) {
+        count++;
+    }
+
+    if (!feof(f)) { // fgets error
+        count = 0;
+    }
+
+    fclose(f);
+    f = NULL;
+    return count;
+}
+
+/* Parse start/end address of all segments from maps file and
+   locate the php .text segment.
+   Return TRUE if succeed; FALSE if fail to parse maps file.
+*/
+static bool parse_map_file(struct map_seg_addresses* map_segments,
+                           size_t map_segments_count,
+                           struct map_seg_addresses* php_text_segment)
+{
+    char buffer[MAXPATHLEN];
+    FILE *f = NULL;
+    f = fopen("/proc/self/maps", "r");
+    if (!f) return false;
+
+    size_t index = 0;
+    uintptr_t start; /* segment start address */
+    uintptr_t end;   /* segment end address */
+    bool ret = true; /* function return value */
+
+    /* Get PHP .text segment lighthouse address used
+       to locate PHP .text segment */
+    uintptr_t php_lighthouse_addr = (uintptr_t) php_text_lighthouse();
+
+    /* Store each line in tha map_segments array */
+    while (fgets(buffer, MAXPATHLEN, f)) {
+        /* get first two fields: start-end address */
+        if (sscanf(buffer, "%lx-%lx", &start, &end) == 2 &&
+            index < map_segments_count)
+        {
+            map_segments[index].start = start;
+            map_segments[index].end = end;
+            index++;
```

Comment:
If you search only space before the ``.text``, it doesn't make sense to store segments with greater addresses. I see ~500 such segments.
Personally, I think the space after ``.text`` also should be considered. So we have to keep information only about close segments. (e.g. started/ended in 4GB around PHP .text segment)

Reply:
Thanks. Since we assume the maps file is not sorted, we store all segments. After combining all segments, there are only a few segments left.
We will consider the space after .text again and check the influence on the heap.

Reply:
We verified that we can use space after [heap] segment. we will update patch and search any candidate BEFORE and AFTER PHP .text segment.